r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java

```diff
@@ -440,6 +444,11 @@ public boolean cancel()
     };
   }
 
+  private boolean isWaiterTimeoutWithinFastFailThreshold()
+  {
+    return _waiterTimeout >= MIN_WAITER_TIMEOUT && _waiterTimeout <= MAX_WAITER_TIMEOUT;
+  }
+
   @Override
   public void put(T obj)
   {
```

Comment:
You can also fix the leak in the SingleTimeout

```
  public SingleTimeout(ScheduledExecutorService executor, long timeout, TimeUnit timeoutUnit, T item, Consumer<T> timeoutAction) 
  { 
    ArgumentUtil.ensureNotNull(item,"item"); 
    ArgumentUtil.ensureNotNull(timeoutAction,"timeoutAction"); 
 
    _item = new AtomicReference<>(item); 
    _future = executor.schedule(() -> { 
      T item1 = _item.getAndSet(null); 
      if (item1 != null) 
      { 
        try 
        { 
          timeoutAction.apply(item1); 
        } catch (Throwable e) 
        { 
          LOG.error("Failed to execute timeout action", e); 
        } 
      } 
    }, timeout, timeoutUnit); 
  } 
```

Reply:
Done

Reply:
SingleTime not part of this PR.

Reply:
You mean - you dont want to include singletimeout? It is already present in this PR as a separate commit following this. I will squash these during merge time